### DISCUSSION TITLE
Problem with Member_2 / Member and sendMessage method

### DISCUSSION THREAD

**Discussion by ShadyManu** [OfficeDev/microsoft-365-agents-toolkit#11694] [Q&A] [answered]
Hello everyone, I'm having some major issues optimizing a notification bot made with Teams Toolkit.

My main problem is that:
    `      const member = await notificationApp.notification.findMember(
        async (m) => m.account.email === email
      );`
takes literally too long. with barely 50 members, it takes 1 minute to execute that, it means that to send 50 single messages it's taking almost one hour.

So I hosted a Redis service on Azure, and I wanted to save the user inside redis, which reduce a lot the time to read the user. The problema is that I'm not able to build the Member from the info I'm saving on Redis. Considering that I can just save json on Redis, I saved what I could and I'm trying to build Member class from it's constructor. This is more or less what I have done:
`
    let memberFromRedis = await getMemberFromRedis(email);
    const account = memberFromRedis["account"] as TeamsChannelAccount;
    const conversationReference = memberFromRedis["parent"]["conversationReference"];
    const teamsBotInstallation = await notificationApp.notification.buildTeamsBotInstallation(conversationReference);
    // @ts-ignore
    const user = new Member(teamsBotInstallation , account);
` 
I'm using ts-ignore because I'm getting this weird type error:
![image](https://github.com/OfficeDev/teams-toolkit/assets/6276930/e499838c-5ea1-4b3e-936a-8acb7bd728b0)
translated it's:
Argument type 'TeamsBotInstallation_2' it is not assignable to parameeter type 'TeamsBotInstallation'. 
    Types of property 'adapter' are not compatible.
        In the type 'CloudAdapter' are missing the following properties of the type 'BotFrameworkAdapter': TokenApiClientCredentialKey, credentials, credentialsProvider, settings and other 41.

but inside the types I can see:
TeamsBotInstallation_2 as TeamsBotInstallation,
so they should be the same type right?

anyway, debugging this code seems the user is the same but when I come to 
    `await user.sendMessage("test");`
I'm getting this error:
![image](https://github.com/OfficeDev/teams-toolkit/assets/6276930/a0fa162e-60ac-4ee3-a610-9b9756e427b8)
`
{
    "code": "Internal",
    "message": "Error: `CloudAdapterBase.continueConversation` is deprecated, please use `CloudAdapterBase.continueConversationAsync`"
}
`

for some reasons, the type coming back from findMember (Member_2) it looks like slighlty different from Member class, even tho in this case as well I can see from types export Member_2 as Member, so I expect them to be exactly the same.

with some breakpoint I can notice just a small difference:
![image](https://github.com/OfficeDev/teams-toolkit/assets/6276930/6073d391-f5ce-4b9c-9430-621c6ba75abb)

the functions (not just this one but sendAdaptiveCard and the others as well) seems to point to different locations, the type of the object seems different.. is this normal or is this the problem? anyway, how can I solve that?

**Comment by qinezh** [comment, marked as answer]
There're two `Member` types in TeamsFx SDK:
1. [(conversation) Member](https://github.com/OfficeDev/teams-toolkit/blob/dev/packages/sdk/src/conversation/notification.ts#L185)
2. [(conversationWithCloudAdapter) Member](https://github.com/OfficeDev/teams-toolkit/blob/dev/packages/sdk/src/conversationWithCloudAdapter/notification.ts#L201)

The first one is actually deprecated as one of its property [(conversation) TeamsBotInstallation](https://github.com/OfficeDev/TeamsFx/blob/dev/packages/sdk/src/conversation/notification.ts#L321) is deprecated and it's recommend to use the new type with the same name [(conversationWithCloudAdapter) TeamsBotInstallation](https://github.com/OfficeDev/TeamsFx/blob/dev/packages/sdk/src/conversationWithCloudAdapter/notification.ts#L345).

There's type error from your code is, type of [(conversationWithCloudAdapter) TeamsBotInstallation](https://github.com/OfficeDev/TeamsFx/blob/dev/packages/sdk/src/conversationWithCloudAdapter/notification.ts#L345) is retuned by `buildTeamsBotInstallation`, while the constructor of [(conversation) Member](https://github.com/OfficeDev/teams-toolkit/blob/dev/packages/sdk/src/conversation/notification.ts#L185) requires type of [(conversation) TeamsBotInstallation](https://github.com/OfficeDev/TeamsFx/blob/dev/packages/sdk/src/conversation/notification.ts#L321)

For your issue, you may take a look at the sample app of [Large Scale Notification Bot](https://github.com/OfficeDev/teams-toolkit-samples/blob/dev/large-scale-notification/README.md) if you have performance concerns about sending notifications.

**Reply by m-raso-hyntelo** [reply]
Thanks! Yes that solved my probem.. exporting both as Member literally confused me, I was using the wrong Member, so I had to specific which one in this one:
`BotBuilderCloudAdapter.Member`
once changed Member, everything seems working in local! thanks!
